File: problems/leetcode/leetcode-438-find-all-anagrams-in-a-string/main.cc
#include <string>
#include <unordered_map>
#include <vector>
using namespace std;
class Solution {
   public:
    vector<int> findAnagrams(string s, string p) {
        unordered_map<char, int> d1;  // p 的字符统计
        unordered_map<char, int> d2;  // 窗口内的字符统计
        for (auto ch : p) d1[ch]++;
        auto check = [&]() {
            for (char ch = 'a'; ch <= 'z'; ch++)
                if (d1[ch] != d2[ch]) return false;
            return true;
        };
        int L = 0;
        vector<int> ans;
        for (int R = 0; R < s.size();R++){
            d2[s[R]]++;
            while (L < R && R - L + 1 > p.size()) d2[s[L++]]--;
            if (check()) ans.push_back(L);
        }
        return ans;
    }
};

File: problems/leetcode/leetcode-438-find-all-anagrams-in-a-string/main.py
from collections import defaultdict
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        d = defaultdict(int)  # p 中的各个字符的出现次数
        a = 0  # p 中不同的字符个数
        for ch in p:
            d[ch] += 1
            if d[ch] == 1:
                a += 1

        b = 0  # b 是窗口内的扣减了 p 中字符的种类数
        ans = []

        L = R = 0

        while R < len(s):
            # 扣减出现频次,
            d[s[R]]-=1
            # 并且, 如果扣减后是 0 , 则记录一次 b
            if d[s[R]] == 0:
                b += 1

            while L < R and R - L + 1 > len(p):
                # 恢复出现频次
                d[s[L]] += 1
                # 如果恢复前是 0, 则也要恢复 b
                if d[s[L]] == 1:
                    b -= 1
                L += 1

            if a == b:
                ans.append(L)

            R += 1
        return ans
